src/coreclr/tools/Common/TypeSystem/MetadataEmitter/TypeSystemMetadataEmitter.cs

```diff
@@ -386,6 +383,51 @@ public void Push()
                 }
             }
 
+            public void EmitArrayShapeAtCurrentIndexStack(BlobBuilder signatureBuilder, int rank)
+            {
+                var shapeEncoder = new ArrayShapeEncoder(signatureBuilder);
+
+                bool emittedWithShape = false;
+
+                if (!Complete)
+                {
+                    if (_embeddedDataIndex < _embeddedData.Length)
+                    {
+                        if (_embeddedData[_embeddedDataIndex].kind == EmbeddedSignatureDataKind.ArrayShape)
+                        {
+                            string indexData = string.Join(".", _indexStack);
+
+                            var arrayShapePossibility = _embeddedData[_embeddedDataIndex].index.Split('|');
+                            if (arrayShapePossibility[0] == indexData)
+                            {
+                                string[] boundsStr = arrayShapePossibility[1].Split(',', StringSplitOptions.RemoveEmptyEntries);
+                                string[] loBoundsStr = arrayShapePossibility[2].Split(',', StringSplitOptions.RemoveEmptyEntries);
+                                int[] bounds = new int[boundsStr.Length];
+                                int[] loBounds = new int[loBoundsStr.Length];
+
+                                for (int i = 0; i < boundsStr.Length; i++)
+                                {
+                                    bounds[i] = Int32.Parse(boundsStr[i]);
+                                }
+                                for (int i = 0; i < loBoundsStr.Length; i++)
+                                {
+                                    loBounds[i] = Int32.Parse(loBoundsStr[i]);
+                                }
+
+                                shapeEncoder.Shape(rank, ImmutableArray.Create(bounds), ImmutableArray.Create(loBounds));
```

Comment:
Nit - would it be useful to add assertions checking the bounds / lower bounds length against the array rank?